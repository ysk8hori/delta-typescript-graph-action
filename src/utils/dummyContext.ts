import { vi } from 'vitest';
import type { Context } from './context';
import type GitHub from './github';

export function getDummyContext(context?: {
  configExclude?: string[];
  configTsconfig?: string;
}): Context {
  return {
    config: {
      tsconfigRoot: '',
      tsconfig: context?.configTsconfig ?? '',
      maxSize: 100,
      orientation: {},
      debugEnabled: false,
      inDetails: false,
      exclude: context?.configExclude ?? ['node_modules'],
      includeIndexFileDependencies: false,
      commentTitle: '',
      showMetrics: false,
    },
    github: {
      getWorkflowName: vi.fn(),
      commentToPR: vi.fn(),
      deleteComment: vi.fn(),
      getTSFiles: vi.fn(),
      getBaseSha: vi.fn(),
      getHeadSha: vi.fn(),
      cloneRepo: vi.fn(),
    } as unknown as GitHub,
    fullCommentTitle: '## Delta TypeScript Graph<!--test-workflow.yml-->',
    filesChanged: {
      created: [],
      deleted: [],
      modified: [],
      renamed: [],
    },
  };
}
