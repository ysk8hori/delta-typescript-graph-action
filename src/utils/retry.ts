export function retry<T>(
  fn: () => Promise<T>,
  maxRetries = 2,
  interval = 1000,
): Promise<T | undefined> {
  return fn().catch(async (error: Error) => {
    console.warn('\n', error);
    if (maxRetries <= 0) return undefined;

    await new Promise(resolve => setTimeout(resolve, interval));
    return retry(fn, maxRetries - 1);
  });
}
