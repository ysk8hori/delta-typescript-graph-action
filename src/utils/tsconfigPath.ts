import path from 'node:path';
import type { Config } from './config';

/**
 * Get the base directory for tsconfig
 */
function getTsconfigBaseDir(config: Pick<Config, 'tsconfig' | 'tsconfigRoot'>): string {
  return config.tsconfig ? path.dirname(config.tsconfig) : config.tsconfigRoot;
}

/**
 * Convert an absolute file path to a path relative to the tsconfig base directory
 * @returns null if the file is outside the tsconfig scope
 */
function getRelativePathFromTsconfig(
  absolutePath: string,
  config: Pick<Config, 'tsconfig' | 'tsconfigRoot'>
): string | null {
  const baseDir = getTsconfigBaseDir(config);
  const relativePath = path.relative(baseDir, absolutePath);
  
  // Files outside tsconfig scope have relative paths starting with '..'
  if (relativePath.startsWith('..')) {
    return null;
  }
  
  return relativePath;
}

/**
 * Check if a file is within the tsconfig scope
 */
export function isInTsconfigScope(
  absolutePath: string,
  config: Pick<Config, 'tsconfig' | 'tsconfigRoot'>
): boolean {
  return getRelativePathFromTsconfig(absolutePath, config) !== null;
}

/**
 * Filter an array of file paths to only include those within the tsconfig scope
 */
export function filterFilesInTsconfigScope(
  filePaths: string[],
  config: Pick<Config, 'tsconfig' | 'tsconfigRoot'>
): string[] {
  return filePaths.filter(filePath => isInTsconfigScope(filePath, config));
}

/**
 * Convert file paths to relative paths from tsconfig base directory
 * Only includes files within the tsconfig scope
 */
export function convertToRelativePathsFromTsconfig(
  filePaths: string[],
  config: Pick<Config, 'tsconfig' | 'tsconfigRoot'>
): string[] {
  return filePaths
    .map(filePath => getRelativePathFromTsconfig(filePath, config))
    .filter((relativePath): relativePath is string => relativePath !== null);
}