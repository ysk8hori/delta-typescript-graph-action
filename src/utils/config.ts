// TODO: ワークフローからパラメータを受け取って使用するようにする

import { z } from 'zod';
import type mermaidify from '@ysk8hori/typescript-graph/dist/src/mermaidify';
import * as core from '@actions/core';
import { log } from './log';

/** tsconfig のルートディレクトリ */
const TSCONFIG_ROOT = 'tsconfig-root';
/** 変更ファイル数が多い場合にグラフの表示を抑止するが、その際のノード数を指定する値 */
const MAX_SIZE = 'max-size';
/** グラフの方向を指定する */
const ORIENTATION = 'orientation';
/** デバッグモード */
const DEBUG = 'debug';
/** Mermaid を `<details>` タグで囲み折りたたむかどうか */
const IN_DETAILS = 'in-details';
/** ファイルの除外対象 */
const EXCLUDE = 'exclude';
/** 変更対象のファイルが同階層の index.ts などから参照されている場合、その index.ts への依存ファイルも表示するかどうか */
const INCLUDE_INDEX_FILE_DEPENDENCIES = 'include-index-file-dependencies';

const tsgConfigScheme = z.object({
  /** tsconfig のルートディレクトリ */
  tsconfigRoot: z.string().optional(),
  /** 変更ファイル数が多い場合にグラフの表示を抑止するが、その際のノード数を指定する値 */
  maxSize: z.number().optional(),
  /** グラフの方向を指定する */
  orientation: z.union([z.literal('TB'), z.literal('LR')]).optional(),
  /** デバッグモード */
  debug: z.boolean().optional(),
  /** Mermaid を `<details>` タグで囲み折りたたむかどうか */
  inDetails: z.boolean().optional(),
  /** ファイルの除外対象 */
  exclude: z.array(z.string()).optional(),
  /** 変更対象のファイルが同階層の index.ts などから参照されている場合、その index.ts への依存ファイルも表示するかどうか */
  includeIndexFileDependencies: z.boolean().optional(),
});

export type TsgConfigScheme = z.infer<typeof tsgConfigScheme>;

export function loggingConfig() {
  // それぞれの値を core.getInput で取得し、取得したままの値をオブジェクトにして JSON にして出力する
  const config = {
    tsconfigRoot: core.getInput(TSCONFIG_ROOT),
    maxSize: core.getInput(MAX_SIZE),
    orientation: core.getInput(ORIENTATION),
    debug: core.getInput(DEBUG),
    inDetails: core.getInput(IN_DETAILS),
    exclude: core.getInput(EXCLUDE),
    includeIndexFileDependencies: core.getInput(
      INCLUDE_INDEX_FILE_DEPENDENCIES,
    ),
  };
  log('config:', JSON.stringify(config, null, 2));
}

/**
 * tsconfig を探索するディレクトリ情報を取得する。
 *
 * ここで指定した階層から上の階層に向かって tsconfig を探索する。
 */
export function getTsconfigRoot(): string {
  return core.getInput(TSCONFIG_ROOT) ?? './';
}

/** 変更ファイル数が多い場合にグラフの表示を抑止するが、その際のノード数を指定する値を取得する。 */
export function getMaxSize(): number {
  const maxSize = core.getInput(MAX_SIZE);
  if (maxSize && !isNaN(parseInt(maxSize, 10))) {
    return parseInt(maxSize, 10);
  }
  return 30;
}

/** グラフの方向を指定したオブジェクトを取得する */
export function getOrientation(): Pick<
  Parameters<typeof mermaidify>[2],
  'LR' | 'TB'
> {
  const orientation = core.getInput(ORIENTATION);
  if (orientation === 'TB') {
    return { TB: true };
  }
  if (orientation === 'LR') {
    return { LR: true };
  }
  return {};
}

export function isDebugEnabled(): boolean {
  return core.getInput(DEBUG) === 'true';
}

/** Mermaid を `<details>` タグで囲み折りたたむかどうか */
export function isInDetails(): boolean {
  return core.getInput(IN_DETAILS) === 'true';
}

export function exclude(): string[] {
  return core
    .getInput(EXCLUDE)
    .split(',')
    .map(s => s.trim());
}

/** 変更対象のファイルが同階層の index.ts などから参照されている場合、その index.ts への依存ファイルも表示するかどうか */
export function isIncludeIndexFileDependencies(): boolean {
  return core.getInput(INCLUDE_INDEX_FILE_DEPENDENCIES) === 'true';
}
