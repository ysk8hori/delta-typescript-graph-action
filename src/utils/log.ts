import { isDebugEnabled } from './config';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function log(message?: any, ...optionalParams: any[]): void {
  if (!isDebugEnabled()) return;
  console.log(stringifyObject(message), ...optionalParams.map(stringifyObject));
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function stringifyObject(message: any) {
  // message がオブジェクトの場合は JSON.stringify で文字列化する あ
  if (message !== null && typeof message === 'object') {
    return JSON.stringify(message, null, 2);
  }
  return message;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function info(message?: any, ...optionalParams: any[]): void {
  console.info(
    stringifyObject(message),
    ...optionalParams.map(stringifyObject),
  );
}
