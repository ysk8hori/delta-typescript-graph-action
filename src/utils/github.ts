import * as core from '@actions/core';
import * as github from '@actions/github';
import { info, log } from './log';
import { execSync } from 'child_process';
import { retry } from './retry';

export type PullRequestFileInfo = {
  filename: string;
  status:
    | 'added'
    | 'removed'
    | 'modified'
    | 'renamed'
    | 'copied'
    | 'changed'
    | 'unchanged';
  previous_filename: string | undefined;
};

export type PullRequestFilesInfo = {
  created: PullRequestFileInfo[];
  deleted: PullRequestFileInfo[];
  modified: PullRequestFileInfo[];
  renamed: PullRequestFileInfo[];
  unchanged: PullRequestFileInfo[];
};

/**
 * 400、401、403、404、422、451を除く、サーバーの4xx/5xx応答の場合はエラーをスローする。
 *
 * @param e エラーオブジェクト
 * @returns 400、401、403、404、422、451 の場合は undefined を返す。
 * @see https://github.com/octokit/plugin-retry.js
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function throwUnexpectedError(e: any) {
  // @see https://github.com/octokit/plugin-retry.js
  if ([400, 401, 403, 404, 422, 451].includes(e?.status)) {
    console.warn(e);
    return undefined;
  }
  throw new Error(e.message, { cause: e });
}

export default class GitHub {
  #octokit: ReturnType<typeof github.getOctokit>;

  constructor() {
    this.#octokit = github.getOctokit(core.getInput('access-token'));
  }

  public async getTSFiles(): Promise<PullRequestFilesInfo> {
    const compareResult =
      await this.#octokit.rest.repos.compareCommitsWithBasehead({
        owner: github.context.repo.owner,
        repo: github.context.repo.repo,
        basehead: `${github.context.payload.pull_request?.base.sha}...${github.context.payload.pull_request?.head.sha}`,
      });
    const files = compareResult.data.files
      ?.filter(file =>
        /\.ts$|\.tsx$|\.vue$|\.astro$|\.svelte$/.test(file.filename),
      )
      .map(file => ({
        filename: file.filename,
        status: file.status,
        previous_filename: file.previous_filename,
      }));
    log(files);

    // typescript-graph では、以下の分類で処理する。
    return {
      created:
        files?.filter(
          file => file.status === 'added' || file.status === 'copied',
        ) ?? [],
      deleted: files?.filter(file => file.status === 'removed') ?? [],
      modified:
        files?.filter(
          file => file.status === 'modified' || file.status === 'changed',
        ) ?? [],
      renamed: files?.filter(file => file.status === 'renamed') ?? [],
      unchanged: files?.filter(file => file.status === 'unchanged') ?? [],
    };
  }

  public getBaseSha() {
    return github.context.payload.pull_request?.base.sha;
  }

  public getHeadSha() {
    return github.context.payload.pull_request?.head.sha;
  }

  /**
   * コメントのタイトルは同一コメントを探す際にも使用する
   */
  public getCommentTitle() {
    return `## Delta TypeScript Graph<!--${github.context.workflow}-->`;
  }

  /**
   * PRにコメントする。
   *
   * 同一PR内では同一コメントを上書きし続ける。
   */
  public async commentToPR(body: string) {
    const owner = github.context.repo.owner;
    const repo = github.context.repo.repo;
    const issue_number = github.context.payload.number;
    github.context.workflow;
    // 1. 既存のコメントを取得する
    const comments = await retry(() =>
      this.#octokit.rest.issues
        .listComments({
          owner,
          repo,
          issue_number,
        })
        .catch(throwUnexpectedError),
    );

    if (!comments) {
      info('commentToPR:コメントの取得に失敗しました');
      return;
    }

    // 2. 既存のコメントがあれば、そのコメントのIDを取得する
    const existingComment = comments.data.find(comment =>
      comment.body?.trim().startsWith(this.getCommentTitle()),
    );

    if (existingComment) {
      // 3. 既存のコメントがあれば、そのコメントを更新する
      await retry(() =>
        this.#octokit.rest.issues
          .updateComment({
            owner,
            repo,
            comment_id: existingComment.id,
            body,
          })
          .catch(throwUnexpectedError),
      );
    } else {
      // 4. 既存のコメントがなければ、新規にコメントを作成する
      await retry(() =>
        this.#octokit.rest.issues
          .createComment({
            owner,
            repo,
            issue_number,
            body,
          })
          .catch(throwUnexpectedError),
      );
    }
  }

  /**
   * PRのコメントを削除する
   */
  public async deleteComment() {
    const owner = github.context.repo.owner;
    const repo = github.context.repo.repo;
    const issue_number = github.context.payload.number;
    // 1. 既存のコメントを取得する
    const comments = await retry(() =>
      this.#octokit.rest.issues
        .listComments({
          owner,
          repo,
          issue_number,
        })
        .catch(throwUnexpectedError),
    );

    if (!comments) {
      info('deleteComment:コメントの取得に失敗しました');
      return;
    }

    // 2. 既存のコメントがあれば、そのコメントのIDを取得する
    const existingComment = comments.data.find(comment =>
      comment.body?.trim().startsWith(this.getCommentTitle()),
    );

    if (existingComment) {
      // 3. 既存のコメントがあれば、そのコメントを削除する
      await retry(() =>
        this.#octokit.rest.issues
          .deleteComment({
            owner,
            repo,
            comment_id: existingComment.id,
          })
          .catch(throwUnexpectedError),
      );
    }
  }

  public cloneRepo() {
    const repo = github.context.repo;
    // リポジトリのURLを取得
    const repoUrl = `https://github.com/${repo.owner}/${repo.repo}.git`;
    // リポジトリをチェックアウト
    execSync(`git clone ${repoUrl}`);
    // result としてリポジトリ名を返す
    return { repoDir: repo.repo };
  }
}
