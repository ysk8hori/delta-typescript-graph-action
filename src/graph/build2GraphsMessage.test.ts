import type { Graph, Node } from '@ysk8hori/typescript-graph';
import { getDummyContext } from '../utils/dummyContext';
import { build2GraphsMessage } from './build2GraphsMessage';
//hoge
const a: Node = {
  path: 'src/A.tsx',
  name: 'A',
  changeStatus: 'not_modified',
};
const b: Node = {
  path: 'src/B.tsx',
  name: 'B',
  changeStatus: 'not_modified',
};
const c: Node = {
  path: 'src/C.tsx',
  name: 'C',
  changeStatus: 'not_modified',
};

const d: Node = {
  path: 'src/1/D.tsx',
  name: 'D',
  changeStatus: 'not_modified',
};

const e: Node = {
  path: 'src/1/E.tsx',
  name: 'E',
  changeStatus: 'not_modified',
};

test('削除がある場合', async () => {
  const base: Graph = {
    nodes: [a, b, c, d, e],
    relations: [
      {
        from: a,
        to: b,
        kind: 'depends_on',
        changeStatus: 'not_modified',
      },
      {
        from: a,
        to: c,
        kind: 'depends_on',
        changeStatus: 'not_modified',
      },
      {
        from: a,
        to: d,
        kind: 'depends_on',
        changeStatus: 'not_modified',
      },
      {
        from: d,
        to: e,
        kind: 'depends_on',
        changeStatus: 'not_modified',
      },
    ],
  };
  const head: Graph = {
    nodes: [a, c, d, e],
    relations: [
      {
        from: a,
        to: c,
        kind: 'depends_on',
        changeStatus: 'not_modified',
      },
      {
        from: a,
        to: d,
        kind: 'depends_on',
        changeStatus: 'not_modified',
      },
      {
        from: d,
        to: e,
        kind: 'depends_on',
        changeStatus: 'not_modified',
      },
    ],
  };
  const context = getDummyContext();
  const result = await build2GraphsMessage(base, head, {
    ...context,
    filesChanged: {
      created: [],
      deleted: [
        { filename: b.path, previous_filename: undefined, status: 'removed' },
      ],
      modified: [
        { filename: a.path, previous_filename: undefined, status: 'modified' },
      ],
      renamed: [],
    },
  });
  expect(result).toMatchSnapshot();
});
