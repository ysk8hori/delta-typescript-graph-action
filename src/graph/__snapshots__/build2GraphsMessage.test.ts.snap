// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`削除がある場合 1`] = `
"


\`\`\`bash
tsg --include src/B.tsx src/A.tsx --highlight src/B.tsx src/A.tsx --exclude node_modules --abstraction src/1
\`\`\`

### Base Branch

\`\`\`mermaid
flowchart
    classDef modified fill:yellow,stroke:#999,color:black
    classDef deleted fill:dimgray,stroke:#999,color:black,stroke-dasharray: 4 4,stroke-width:2px;
    subgraph src["src"]
        src/A.tsx["A"]:::modified
        src/B.tsx["B"]:::deleted
        src/C.tsx["C"]
        src/1["/1"]:::dir
    end
    src/A.tsx-->src/B.tsx
    src/A.tsx-->src/C.tsx
    src/A.tsx-->src/1

\`\`\`

### Head Branch

\`\`\`mermaid
flowchart
    classDef modified fill:yellow,stroke:#999,color:black
    subgraph src["src"]
        src/A.tsx["A"]:::modified
        src/C.tsx["C"]
        src/1["/1"]:::dir
    end
    src/A.tsx-->src/C.tsx
    src/A.tsx-->src/1

\`\`\`



"
`;
